janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMultiQueryStrategy.java

```diff
@@ -0,0 +1,117 @@
+
+/**
+ * @author Marko A. Rodriguez (https://markorodriguez.com)
+ * @author Matthias Broecheler (http://matthiasb.com)
+ */
+public class JanusGraphMultiQueryStrategy extends AbstractTraversalStrategy<TraversalStrategy.ProviderOptimizationStrategy> implements TraversalStrategy.ProviderOptimizationStrategy {
+
+    private static final Set<Class<? extends ProviderOptimizationStrategy>> PRIORS = Collections.singleton(JanusGraphLocalQueryOptimizerStrategy.class);
+    private static final JanusGraphMultiQueryStrategy INSTANCE = new JanusGraphMultiQueryStrategy();
+
+    private JanusGraphMultiQueryStrategy() {
+    }
+
+    @Override
+    public void apply(final Admin<?, ?> traversal) {
+        if (!traversal.getGraph().isPresent()
+            || TraversalHelper.onGraphComputer(traversal)
+            // The LazyBarrierStrategy is not allowed to run on traversals which use drop(). As a precaution,
+            // this strategy should not run on those traversals either, because it can also insert barrier().
+            || !TraversalHelper.getStepsOfAssignableClassRecursively(DropStep.class, traversal).isEmpty()) {
+            return;
+        }
+
+        final Graph graph = traversal.getGraph().get();
+        final StandardJanusGraph janusGraph = graph instanceof StandardJanusGraphTx ? ((StandardJanusGraphTx) graph).getGraph() : (StandardJanusGraph) graph;
+        if (!janusGraph.getConfiguration().useMultiQuery()) return;
+
+        insertMultiQuerySteps(traversal, janusGraph.getConfiguration().limitBatchSize());
+        configureMultiQueriables(traversal);
+    }
+
+    /**
+     * Insert JanusGraphMultiQuerySteps everywhere in the current traversal where MultiQueriable steps could benefit
+     * @param traversal The local traversal layer.
+     */
+    private void insertMultiQuerySteps(final Admin<?,?> traversal, boolean limitBatchSize) {
+        JanusGraphTraversalUtil.getSteps(JanusGraphTraversalUtil::isMultiQueryCompatibleStep, traversal).forEach(step -> {
+            Optional<Step> multiQueryPosition = JanusGraphTraversalUtil.getLocalMultiQueryPositionForStep(step);
+            if (multiQueryPosition.isPresent() && JanusGraphTraversalUtil.isLegalMultiQueryPosition(multiQueryPosition.get())) {
+                Step pos = multiQueryPosition.get();
+                if (limitBatchSize && !(multiQueryPosition.get() instanceof NoOpBarrierStep)) {
+                    NoOpBarrierStep barrier = new NoOpBarrierStep(traversal);
+                    TraversalHelper.insertBeforeStep(barrier, pos, traversal);
+                    pos = barrier;
+                }
+                JanusGraphMultiQueryStep multiQueryStep = new JanusGraphMultiQueryStep(traversal, limitBatchSize);
+                TraversalHelper.insertBeforeStep(multiQueryStep, pos, traversal);
+            }
+        });
+    }
+
+    /**
+     * At this point, all parent traversals have already passed this strategy. Therefore, all JanusGraphMultiQuerySteps
```

Comment:
From your comment, I presume this method may need modifications once https://github.com/JanusGraph/janusgraph/pull/2619 is merged, right?

Reply:
I think so. Having #2619 merged should make thinks easier, but I would have to dig into this part of the code again for more details for sure.

Reply:
After looking into it, I think this should still work even without further action. However, a little bit of refactoring after #2619 is merged would make the code more understandable in some aspects.